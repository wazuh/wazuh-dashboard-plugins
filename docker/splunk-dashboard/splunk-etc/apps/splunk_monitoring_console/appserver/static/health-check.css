.dmc-kpi-item .alert {
    margin-bottom: 0;
}

.dmc-kpi-item h5 {
    font-size: 12px;
}

.dmc-kpi-item .alert-success .icon-alert {
    color: #9ac23c;
}

.dmc-kpi-item .alert-success .icon-alert:before {
    content: '\ECD3';
}

