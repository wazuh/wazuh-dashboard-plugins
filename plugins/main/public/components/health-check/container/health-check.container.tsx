/*
 * Wazuh app - Health Check Component
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import {
  EuiButton,
  EuiCallOut,
  EuiDescriptionList,
  EuiSpacer,
} from '@elastic/eui';
import React, { Fragment, useState, useEffect, useRef } from 'react';
import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
import { AppState, ErrorHandler } from '../../../react-services';
import { useAppConfig, useRootScope } from '../../../components/common/hooks';
import {
  checkApiService,
  checkIndexPatternService,
  checkPatternSupportService,
  checkSetupService,
} from '../services';
import { CheckResult } from '../components/check-result';
import { withErrorBoundary, withReduxProvider } from '../../common/hocs';
import { getCore, getHttp, getWzCurrentAppID } from '../../../kibana-services';
import {
  HEALTH_CHECK_REDIRECTION_TIME,
  NOT_TIME_FIELD_NAME_INDEX_PATTERN,
  WAZUH_INDEX_TYPE_MONITORING,
  WAZUH_INDEX_TYPE_STATISTICS,
  WAZUH_INDEX_TYPE_VULNERABILITIES,
  WAZUH_INDEX_TYPE_FIM,
} from '../../../../common/constants';

import { compose } from 'redux';
import { getThemeAssetURL, getAssetURL } from '../../../utils/assets';
import { serverApis } from '../../../utils/applications';
import { RedirectAppLinks } from '../../../../../../src/plugins/opensearch_dashboards_react/public';

const checks = {
  api: {
    title: 'Check Wazuh API connection',
    label: 'API connection',
    validator: checkApiService,
    awaitFor: [],
    canRetry: true,
  },
  setup: {
    title: 'Check Wazuh API version',
    label: 'API version',
    validator: checkSetupService,
    awaitFor: ['api'],
  },
  pattern: {
    title: 'Check alerts index pattern',
    label: 'Alerts index pattern',
    validator: checkIndexPatternService,
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  patternMonitoring: {
    title: 'Check monitoring index pattern',
    label: 'Monitoring index pattern',
    validator: appConfig =>
      checkPatternSupportService(
        appConfig.data['wazuh.monitoring.pattern'],
        WAZUH_INDEX_TYPE_MONITORING,
      ),
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  patternStatistics: {
    title: 'Check statistics index pattern',
    label: 'Statistics index pattern',
    validator: appConfig =>
      checkPatternSupportService(
        `${appConfig.data['cron.prefix']}-${appConfig.data['cron.statistics.index.name']}-*`,
        WAZUH_INDEX_TYPE_STATISTICS,
      ),
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  'vulnerabilities.pattern': {
    title: 'Check vulnerabilities index pattern',
    label: 'Vulnerabilities index pattern',
    validator: appConfig =>
      checkPatternSupportService(
        appConfig.data['vulnerabilities.pattern'],
        WAZUH_INDEX_TYPE_VULNERABILITIES,
        NOT_TIME_FIELD_NAME_INDEX_PATTERN,
      ),
    awaitFor: [],
    shouldCheck: false,
    canRetry: true,
  },
  'fim.pattern': {
    title: 'Check fim index pattern',
    label: 'Fim index pattern',
    validator: appConfig =>
      checkPatternSupportService(
        appConfig.data['fim.pattern'],
        WAZUH_INDEX_TYPE_FIM,
        NOT_TIME_FIELD_NAME_INDEX_PATTERN,
      ),
    awaitFor: [],
    shouldCheck: false,
    canRetry: true,
  },
};

function HealthCheckComponent() {
  const [checkWarnings, setCheckWarnings] = useState<{ [key: string]: [] }>({});
  const [checkErrors, setCheckErrors] = useState<{ [key: string]: [] }>({});
  const [checksReady, setChecksReady] = useState<{ [key: string]: boolean }>(
    {},
  );
  const [isDebugMode, setIsDebugMode] = useState<boolean>(false);
  const appConfig = useAppConfig();
  const checksInitiated = useRef(false);
  const $rootScope = useRootScope();

  const redirectionPassHealthcheck = () => {
    const params = $rootScope.previousParams || {};
    const queryString = Object.keys(params)
      .map(key => key + '=' + params[key])
      .join('&');
    const url =
      `/app/${getWzCurrentAppID()}#` +
      ($rootScope.previousLocation || '') +
      '?' +
      queryString;
    window.location.href = getHttp().basePath.prepend(url);
  };

  const thereAreErrors = Object.keys(checkErrors).length > 0;
  const thereAreWarnings = Object.keys(checkWarnings).length > 0;

  useEffect(() => {
    if (appConfig.isReady && !checksInitiated.current) {
      checksInitiated.current = true;
      AppState.setPatternSelector(appConfig.data['ip.selector']);
    }
  }, [appConfig]);

  useEffect(() => {
    // Redirect to app when all checks are ready
    Object.keys(checks).every(check => checksReady[check]) &&
      !isDebugMode &&
      !thereAreWarnings &&
      (() =>
        setTimeout(
          redirectionPassHealthcheck,
          HEALTH_CHECK_REDIRECTION_TIME,
        ))();
  }, [checksReady]);

  useEffect(() => {
    // Check if Health should not redirect automatically (Debug mode)
    setIsDebugMode(window.location.href.includes('debug'));
  }, []);

  const handleWarnings = (checkID, warnings, parsed) => {
    const newWarnings = parsed
      ? warnings.map(warning =>
          ErrorHandler.handle(warning, 'Health Check', {
            warning: true,
            silent: true,
          }),
        )
      : warnings;
    setCheckWarnings(prev => ({ ...prev, [checkID]: newWarnings }));
  };

  const handleErrors = (checkID, errors, parsed) => {
    const newErrors = parsed
      ? errors.map(error =>
          ErrorHandler.handle(error, 'Health Check', {
            warning: false,
            silent: true,
          }),
        )
      : errors;
    setCheckErrors(prev => ({ ...prev, [checkID]: newErrors }));
  };

  const cleanWarnings = (checkID: string) => {
    delete checkWarnings[checkID];
    setCheckWarnings({ ...checkWarnings });
  };

  const cleanErrors = (checkID: string) => {
    delete checkErrors[checkID];
    setCheckErrors({ ...checkErrors });
  };

  const handleCheckReady = (checkID, isReady) => {
    setChecksReady(prev => ({ ...prev, [checkID]: isReady }));
  };

  const logoUrl = getHttp().basePath.prepend(
    appConfig.data['customization.enabled'] &&
      appConfig.data['customization.logo.healthcheck']
      ? getAssetURL(appConfig.data['customization.logo.healthcheck'])
      : getThemeAssetURL('logo.svg'),
  );

  const renderChecks = () => {
    const showLogButton = thereAreErrors || thereAreWarnings || isDebugMode;
    return Object.keys(checks).map((check, index) => {
      return (
        <CheckResult
          showLogButton={showLogButton}
          key={`health_check_check_${check}`}
          name={check}
          title={checks[check].title}
          awaitFor={checks[check].awaitFor}
          shouldCheck={
            checks[check].shouldCheck || appConfig.data[`checks.${check}`]
          }
          validationService={checks[check].validator(appConfig)}
          handleWarnings={handleWarnings}
          handleErrors={handleErrors}
          cleanWarnings={cleanWarnings}
          cleanErrors={cleanErrors}
          isLoading={appConfig.isLoading}
          handleCheckReady={handleCheckReady}
          checksReady={checksReady}
          canRetry={checks[check].canRetry}
        />
      );
    });
  };

  const addTagsToUrl = error => {
    const words = error.split(' ');
    words.forEach((word, index) => {
      if (word.includes('http://') || word.includes('https://')) {
        if (words[index - 1] === 'guide:') {
          if (word.endsWith('.') || word.endsWith(',')) {
            words[index - 2] = `<a href="${word.slice(
              0,
              -1,
            )}" rel="noopener noreferrer" target="_blank">${
              words[index - 2]
            } ${words[index - 1].slice(0, -1)}</a>${word.slice(-1)}`;
          } else {
            words[
              index - 2
            ] = `<a href="${word}" rel="noopener noreferrer" target="_blank">${
              words[index - 2]
            } ${words[index - 1].slice(0, -1)}</a> `;
          }
          words.splice(index - 1, 2);
        } else {
          if (word.endsWith('.') || word.endsWith(',')) {
            words[index] = `<a href="${word.slice(
              0,
              -1,
            )}" rel="noopener noreferrer" target="_blank">${word.slice(
              0,
              -1,
            )}</a>${word.slice(-1)}`;
          } else {
            words[
              index
            ] = `<a href="${word}" rel="noopener noreferrer" target="_blank">${word}</a>`;
          }
        }
      }
    });
    return words.join(' ');
  };

  const renderWarnings = () => {
    return Object.keys(checkWarnings).map(checkID =>
      checkWarnings[checkID].map((warning, index) => (
        <Fragment key={index}>
          <EuiCallOut
            title={
              <>
                {`[${checks[checkID].label}]`}{' '}
                <span
                  dangerouslySetInnerHTML={{ __html: addTagsToUrl(warning) }}
                ></span>
              </>
            }
            color='warning'
            iconType='alert'
            style={{ textAlign: 'left' }}
            data-test-subj='callOutWarning'
          ></EuiCallOut>
          <EuiSpacer size='xs' />
        </Fragment>
      )),
    );
  };

  const renderErrors = () => {
    return Object.keys(checkErrors).map(checkID =>
      checkErrors[checkID].map((error, index) => (
        <Fragment key={index}>
          <EuiCallOut
            title={
              <>
                {`[${checks[checkID].label}]`}{' '}
                <span
                  dangerouslySetInnerHTML={{ __html: addTagsToUrl(error) }}
                ></span>
              </>
            }
            color='danger'
            iconType='alert'
            style={{ textAlign: 'left' }}
            data-test-subj='callOutError'
          ></EuiCallOut>
          <EuiSpacer size='xs' />
        </Fragment>
      )),
    );
  };

  return (
    <div className='health-check'>
      <img src={logoUrl} className='health-check-logo' alt=''></img>
      <div className='margin-top-30'>
        <EuiDescriptionList textStyle='reverse' align='center' type='column'>
          {renderChecks()}
        </EuiDescriptionList>
      </div>
      {thereAreErrors && (
        <>
          <EuiSpacer size='xl' />
          {renderErrors()}
        </>
      )}
      {thereAreWarnings && (
        <>
          <EuiSpacer size='xl' />
          {renderWarnings()}
        </>
      )}
      {(thereAreErrors || thereAreWarnings || isDebugMode) && (
        <>
          <EuiSpacer size='xl' />
          <EuiFlexGroup justifyContent='center'>
            {thereAreErrors && (
              <EuiFlexItem grow={false}>
                <RedirectAppLinks application={getCore().application}>
                  <EuiButton
                    fill
                    href={getCore().application.getUrlForApp(serverApis.id)}
                  >
                    Go to Settings
                  </EuiButton>
                </RedirectAppLinks>
              </EuiFlexItem>
            )}
            {(isDebugMode || thereAreWarnings) &&
              Object.keys(checks).every(check => checksReady[check]) && (
                <EuiFlexItem grow={false}>
                  <EuiButton fill onClick={redirectionPassHealthcheck}>
                    Continue
                  </EuiButton>
                </EuiFlexItem>
              )}
          </EuiFlexGroup>
        </>
      )}
      <EuiSpacer size='xl' />
    </div>
  );
}

export const HealthCheck = compose(
  withErrorBoundary,
  withReduxProvider,
)(HealthCheckComponent);

export const HealthCheckTest = HealthCheckComponent;
