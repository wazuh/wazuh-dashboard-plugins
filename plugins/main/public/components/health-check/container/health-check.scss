.health-check {
    .euiCodeBlock.euiCodeBlock--hasControl {
        min-height: 0px;
    }
}