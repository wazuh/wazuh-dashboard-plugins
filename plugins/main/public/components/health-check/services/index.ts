/*
 * Wazuh app - Health check services
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

export * from './check-api.service';
export * from './check-plugin-platform-settings.service';
export * from './check-index-pattern/check-index-pattern.service';
export * from './check-pattern-support.service';
export * from './check-setup.service';
