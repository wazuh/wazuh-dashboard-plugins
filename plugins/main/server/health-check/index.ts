export * from './index-patterns';
export * from './templates';
export * from './server-api';
