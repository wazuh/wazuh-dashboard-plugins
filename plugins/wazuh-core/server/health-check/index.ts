export * from './index-patterns';
export * from './templates';
