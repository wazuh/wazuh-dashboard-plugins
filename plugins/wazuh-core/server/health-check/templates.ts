import { InitializationTaskRunContext } from '../services';

export function getTemplateForIndexPattern(
  indexPatternTitle: string,
  templates: { name: string; index_patterns: string }[],
) {
  return templates.filter(
    ({ index_patterns: indexPatternsTemplate }: { index_patterns: string }) => {
      const [, cleanIndexPatterns] = indexPatternsTemplate.match(/\[(.+)]/) || [
        null,
        null,
      ];

      if (!cleanIndexPatterns) {
        return false;
      }

      const indexPatterns = cleanIndexPatterns.match(/([^\s,]+)/g);

      if (!indexPatterns) {
        return false;
      }

      const lastChar = indexPatternTitle.at(-1);
      const indexPatternTitleCleaned =
        lastChar === '*' ? indexPatternTitle.slice(0, -1) : indexPatternTitle;

      return indexPatterns.some(indexPattern => {
        const lastChar = indexPattern.at(-1);
        const indexPatternCleaned =
          lastChar === '*' ? indexPattern.slice(0, -1) : indexPattern;

        return (
          indexPatternCleaned.includes(indexPatternTitleCleaned) ||
          indexPatternTitleCleaned.includes(indexPatternCleaned)
        );
      });
    },
  );
}

export const checkIndexPatternHasTemplate = async (
  { logger }: InitializationTaskRunContext,
  {
    indexPatternTitle,
    opensearchClient,
  }: { indexPatternTitle: string; opensearchClient: any },
) => {
  logger.debug('Getting templates');

  const data = await opensearchClient.cat.templates({ format: 'json' });

  logger.debug(
    'Checking the index pattern with title [${indexPatternTitle}] has defined some template',
  );

  const templatesFound = getTemplateForIndexPattern(
    indexPatternTitle,
    data.body,
  );

  if (templatesFound.length === 0) {
    throw new Error(
      `No template found for index pattern with title [${indexPatternTitle}]`,
    );
  }

  logger.info(
    `Template [${templatesFound
      .map(({ name }) => name)
      .join(
        ', ',
      )}] found for index pattern with title [${indexPatternTitle}]: `,
  );
};

export const initializationTaskCreatorExistTemplate = ({
  getOpenSearchClient,
  getIndexPatternTitle,
  taskName,
}: {
  getOpenSearchClient: (ctx: InitializationTaskRunContext) => any;
  getIndexPatternTitle: (ctx: InitializationTaskRunContext) => Promise<string>;
  taskName: string;
}) => ({
  name: taskName,
  async run(ctx: InitializationTaskRunContext) {
    let indexPatternTitle;

    try {
      ctx.logger.debug('Starting check of existent template');

      const opensearchClient = getOpenSearchClient(ctx);

      indexPatternTitle = await getIndexPatternTitle(ctx);
      await checkIndexPatternHasTemplate(ctx, {
        opensearchClient,
        indexPatternTitle,
      });
      ctx.logger.info('Start check of existent template finished');
    } catch (error) {
      const message = `Error checking of existent template for index pattern with title [${indexPatternTitle}]: ${error.message}`;

      ctx.logger.error(message);
      throw new Error(message);
    }
  },
});
