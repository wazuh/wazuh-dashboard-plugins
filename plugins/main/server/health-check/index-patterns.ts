import { IndexPatternsFetcher } from '../../../../src/plugins/data/server';
import {
  InitializationTaskContext,
  InitializationTaskRunContext,
} from '../services';

interface EnsureIndexPatternExistenceContextTask {
  indexPatternID: string;
  options: any;
}

interface EnsureIndexPatternExistenceContextTaskWithConfigurationSetting
  extends EnsureIndexPatternExistenceContextTask {
  configurationSettingKey: string;
}

// TODO: unused
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const decoratorCheckIsEnabled =
  (
    callback: (
      ctx: InitializationTaskRunContext,
      ctxTask: EnsureIndexPatternExistenceContextTask,
    ) => Promise<void>,
  ) =>
  async (
    ctx: InitializationTaskRunContext,
    {
      configurationSettingKey,
      ...ctxTask
    }: EnsureIndexPatternExistenceContextTaskWithConfigurationSetting,
  ) => {
    if (await ctx.configuration.get(configurationSettingKey)) {
      await callback(ctx, ctxTask);
    } else {
      ctx.logger.info(`Check [${configurationSettingKey}]: disabled. Skipped.`);
    }
  };

async function getFieldMappings(
  { logger, indexPatternsClient },
  indexPatternTitle: string,
) {
  logger.debug(`Getting index pattern fields for title [${indexPatternTitle}]`);

  // https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2.16.0/src/plugins/data/server/index_patterns/routes.ts#L74
  const fields = await indexPatternsClient.getFieldsForWildcard({
    pattern: indexPatternTitle,
    // meta_fields=_source&meta_fields=_id&meta_fields=_type&meta_fields=_index&meta_fields=_score
    metaFields: ['_source', '_id', '_type', '_index', '_score'],
  });

  logger.debug(
    `Fields for index pattern with title [${indexPatternTitle}]: ${JSON.stringify(
      fields,
    )}`,
  );

  return fields;
}

async function createIndexPattern(
  { logger, savedObjectsClient, indexPatternsClient },
  indexPatternID,
  options: {
    fieldsNoIndices?: any;
    savedObjectOverwrite?: Record<string, any>;
  } = {},
) {
  try {
    let fieldsObj;

    try {
      fieldsObj = await getFieldMappings(
        { logger, indexPatternsClient },
        indexPatternID,
      );
    } catch (error) {
      if (error?.output?.statusCode === 404 && options.fieldsNoIndices) {
        const message = `Fields for index pattern with ID [${indexPatternID}] could not be obtained. This could indicate there are not matching indices because they were not generated or there is some error in the process that generates and indexes that data. The index pattern will be created with a set of pre-defined fields.`;

        logger.warn(message);
        fieldsObj = options.fieldsNoIndices;
      } else {
        throw error;
      }
    }

    const title = indexPatternID;
    const fields = JSON.stringify(fieldsObj);

    let savedObjectData = {
      title,
      fields,
    };

    if (typeof options?.savedObjectOverwrite === 'function') {
      const overwriteProps = options?.savedObjectOverwrite(savedObjectData);
      savedObjectData = {
        ...savedObjectData,
        ...overwriteProps,
      };
    }

    if (typeof options?.savedObjectOverwrite === 'object') {
      savedObjectData = {
        ...savedObjectData,
        ...options?.savedObjectOverwrite,
      };
    }

    logger.debug(
      `Creating index pattern with ID [${indexPatternID}] title [${savedObjectData.title}]`,
    );

    const response = await savedObjectsClient.create(
      'index-pattern',
      savedObjectData,
      {
        id: indexPatternID,
        overwrite: true,
        refresh: true,
      },
    );
    const indexPatternCreatedMessage = `Created index pattern with ID [${response.id}] title [${response.attributes.title}]`;

    logger.info(indexPatternCreatedMessage);

    return response;
  } catch (error) {
    throw new Error(
      `index pattern with ID [${indexPatternID}] could not be created due to: ${error.message}. This could indicate the collection is disabled or there is a problem in the data collection or ingestion.`,
    );
  }
}

export const ensureIndexPatternExistence = async (
  { logger, savedObjectsClient, indexPatternsClient },
  { indexPatternID, options = {} }: EnsureIndexPatternExistenceContextTask,
) => {
  try {
    logger.debug(
      `Checking existence of index pattern with ID [${indexPatternID}]`,
    );

    const response = await savedObjectsClient.get(
      'index-pattern',
      indexPatternID,
    );

    logger.debug(`Index pattern with ID [${indexPatternID}] exists`);

    return response;
  } catch (error) {
    // Get not found saved object
    if (error?.output?.statusCode === 404) {
      // Create index pattern
      logger.info(`Index pattern with ID [${indexPatternID}] does not exist`);

      return await createIndexPattern(
        { logger, savedObjectsClient, indexPatternsClient },
        indexPatternID,
        options,
      );
    } else {
      throw new Error(
        `index pattern with ID [${indexPatternID}] existence could not be checked due to: ${error.message}`,
      );
    }
  }
};

export const matchesPatternList = (listString, indexPatternTitle) => {
  const [, cleanIndexPatterns] = listString.match(/\[(.+)]/) || [null, null];

  if (!cleanIndexPatterns) {
    return false;
  }

  const indexPatterns = cleanIndexPatterns.match(/([^\s,]+)/g);

  if (!indexPatterns) {
    return false;
  }

  const lastChar = indexPatternTitle.at(-1);
  const indexPatternTitleCleaned =
    lastChar === '*' ? indexPatternTitle.slice(0, -1) : indexPatternTitle;

  return indexPatterns.some(indexPattern => {
    const lastChar = indexPattern.at(-1);
    const indexPatternCleaned =
      lastChar === '*' ? indexPattern.slice(0, -1) : indexPattern;

    return (
      indexPatternCleaned.startsWith(indexPatternTitleCleaned) ||
      indexPatternTitleCleaned.startsWith(indexPatternCleaned)
    );
  });
};

export const ensureIndexPatternHasTemplate = async (
  ctx,
  indexPatternTitle: string,
) => {
  ctx.logger.debug('Getting templates');
  const data =
    await ctx.services.core.opensearch.client.asInternalUser.cat.templates({
      format: 'json',
    });

  const templates = data.body;

  ctx.logger.debug(`Templates: [${templates}]`);

  const templatesFound = templates.filter(template => {
    if (!template?.index_patterns) {
      return false;
    }

    return matchesPatternList(template?.index_patterns, indexPatternTitle);
  });

  if (templatesFound.length === 0) {
    throw new Error(`Template was not found for [${indexPatternTitle}]`);
  }

  ctx.logger.info(
    `Templates found [${templatesFound.map(({ name }) => name).join(', ')}]`,
  );

  return templatesFound;
};

function getSavedObjectsClient(
  ctx: InitializationTaskRunContext,
  scope: InitializationTaskContext,
) {
  switch (scope) {
    case 'internal': {
      return ctx.services.core.savedObjects.createInternalRepository();
    }

    case 'user': {
      return ctx.services.core.savedObjects.savedObjectsStart.getScopedClient(
        ctx.request,
      );
    }

    default: {
      break;
    }
  }
}

function getIndexPatternsClient(
  ctx: InitializationTaskRunContext,
  scope: InitializationTaskContext,
) {
  switch (scope) {
    case 'internal': {
      return new IndexPatternsFetcher(
        ctx.services.core.opensearch.legacy.client.callAsInternalUser,
      );
    }

    case 'user': {
      return new IndexPatternsFetcher(
        ctx.services.core.opensearch.legacy.client.callAsCurrentUser,
      );
    }

    default: {
      break;
    }
  }
}

async function getIndexPatternID(
  services: any,
  ctx: InitializationTaskRunContext,
  indexPatternID: string,
  configurationSettingKey: string,
) {
  if (indexPatternID) {
    return indexPatternID;
  }
  switch (ctx.scope) {
    case 'internal': {
      return await services.configuration.get(configurationSettingKey);
    }

    case 'user': {
      // TODO
      return ctx.getIndexPatternID(ctx);
    }

    default: {
      break;
    }
  }
}

export const initializationTaskCreatorIndexPattern = ({
  taskName,
  options = {},
  configurationSettingKey,
  indexPatternID,
  services,
  taskMeta = {},
  ...rest
}: {
  taskName: string;
  options: object;
  configurationSettingKey: string;
  indexPatternID?: string;
}) => ({
  isCritical: taskMeta?.isCritical || false,
  name: taskName,
  async run({ context: ctx, logger }: InitializationTaskRunContext) {
    let indexPatternIDResolved;

    try {
      logger.debug('Starting index pattern saved object');
      indexPatternIDResolved = await getIndexPatternID(
        services,
        ctx,
        indexPatternID,
        configurationSettingKey,
      );

      // Get clients depending on the scope
      const savedObjectsClient = getSavedObjectsClient(ctx, ctx.scope);
      const indexPatternsClient = getIndexPatternsClient(ctx, ctx.scope);

      const indexPattern = await ensureIndexPatternExistence(
        { ...ctx, indexPatternsClient, savedObjectsClient, logger },
        {
          indexPatternID: indexPatternIDResolved,
          options,
          configurationSettingKey,
        },
      );

      if (options.hasTemplate) {
        await ensureIndexPatternHasTemplate(
          { ...ctx, logger },
          indexPattern?.attributes?.title,
        );
      }

      return indexPattern;
    } catch (error) {
      const message = `Error initilizating index pattern with ID [${indexPatternIDResolved}]: ${error.message}`;

      logger.error(message);
      throw new Error(message);
    }
  },
});

const fieldMappers = {
  bytes: ({ type }) => (type === 'number' ? { id: 'bytes' } : undefined),
  // integer, remove thousand and decimal separators through the params.pattern
  integer: ({ type }) =>
    type === 'number' ? { id: 'number', params: { pattern: '0' } } : undefined,
  percent: ({ type }) => {
    return type === 'number'
      ? { id: 'percent', params: { pattern: '0,0.[00]%' } }
      : undefined;
  },
};

export function mapFieldsFormat(expectedFields: {
  [key: keyof typeof fieldMappers]: (field: any) => any;
}) {
  return ({ fields }) => {
    const fieldsToMap = Object.keys(expectedFields);
    const mappedFields = JSON.parse(fields)
      ?.filter(({ name }) => fieldsToMap.includes(name))
      .map(field => {
        const { name } = field;
        const mapper = fieldMappers[expectedFields[name]] || undefined;

        if (!mapper) {
          return undefined;
        }
        const result = mapper(field);
        if (!result) {
          return undefined;
        }
        return [name, result];
      })
      .filter(Boolean);

    if (mappedFields.length) {
      return {
        fieldFormatMap: JSON.stringify(Object.fromEntries(mappedFields)),
      }; // Add format map for expected fields
    }
    return {};
  };
}
