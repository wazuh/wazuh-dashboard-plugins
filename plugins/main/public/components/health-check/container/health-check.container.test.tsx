/*
 * Wazuh app - Health Check Component - Test
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { HealthCheckTest } from './health-check.container';

jest.mock('../../../components/common/hooks', () => ({
  useAppConfig: () => ({
    isReady: true,
    isLoading: false,
    data: {
      'ip.selector': 1,
      'checks.metaFields': true,
      'checks.timerFilter': true,
      'checks.maxBuckets': true,
      'checks.api': true,
      'checks.setup': true,
      'checks.pattern': true,
      'checks.template': true,
      'checks.fields': true,
    },
  }),
  useRootScope: () => ({}),
}));

jest.mock('../services', () => ({
  checkPatternService: (appInfo) => () => undefined,
  checkTemplateService: (appInfo) => () => undefined,
  checkApiService: (appInfo) => () => undefined,
  checkSetupService: (appInfo) => () => undefined,
  checkFieldsService: (appInfo) => () => undefined,
  checkPluginPlatformSettings: (appInfo) => () => undefined,
  checkPatternSupportService: (appInfo) => () => undefined,
  checkIndexPatternService: (appInfo) => () => undefined,
}));

jest.mock('../components/check-result', () => ({
  CheckResult: () => () => <></>,
}));

jest.mock('../../../react-services', () => ({
  AppState: {
    setPatternSelector: () => {},
  },
  ErrorHandler: {
    handle: (error) => error,
  },
}));

jest.mock('../../../kibana-services', () => ({
  getHttp: () => ({
    basePath: {
      prepend: (str) => str,
    },
  }),
  getDataPlugin: () => ({
    query: {
      timefilter: {
        timefilter: {
          setTime: (time: number) => true,
        },
      },
    },
  }),
  getUiSettings: () => ({
    get: (setting: string): any => {
      if(setting === 'theme:darkMode'){
        return false
      }
    }
  })
}));

describe('Health Check container', () => {
  it('should render a Health check screen', () => {
    const component = shallow(<HealthCheckTest />);

    expect(component).toMatchSnapshot();
  });

  it('should render a Health check screen with error', () => {
    const component = mount(<HealthCheckTest />);

    component.find('CheckResult').at(1).invoke('handleErrors')('setup', ['Test error']); // invoke is wrapped with act to await for setState

    const callOutError = component.find('EuiCallOut');
    expect(callOutError.text()).toBe('[API version] Test error');
  });
});
