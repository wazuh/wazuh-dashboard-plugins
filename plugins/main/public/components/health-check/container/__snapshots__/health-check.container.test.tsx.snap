// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Health Check container should render a Health check screen 1`] = `
<div
  className="health-check"
>
  <img
    alt=""
    className="health-check-logo"
    src="/plugins/wazuh/assets/images/themes/light/logo.svg"
  />
  <div
    className="margin-top-30"
  >
    <EuiDescriptionList
      align="center"
      textStyle="reverse"
      type="column"
    >
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_api"
        name="api"
        shouldCheck={true}
        showLogButton={false}
        title="Check Wazuh API connection"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={
          Array [
            "api",
          ]
        }
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_setup"
        name="setup"
        shouldCheck={true}
        showLogButton={false}
        title="Check Wazuh API version"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_pattern"
        name="pattern"
        shouldCheck={true}
        showLogButton={false}
        title="Check alerts index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_patternMonitoring"
        name="patternMonitoring"
        shouldCheck={true}
        showLogButton={false}
        title="Check monitoring index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_patternStatistics"
        name="patternStatistics"
        shouldCheck={true}
        showLogButton={false}
        title="Check statistics index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_vulnerabilities.pattern"
        name="vulnerabilities.pattern"
        shouldCheck={true}
        showLogButton={false}
        title="Check vulnerabilities index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        cleanWarnings={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        handleWarnings={[Function]}
        isLoading={false}
        key="health_check_check_fim.pattern"
        name="fim.pattern"
        shouldCheck={true}
        showLogButton={false}
        title="Check fim index pattern"
        validationService={[Function]}
      />
    </EuiDescriptionList>
  </div>
  <EuiSpacer
    size="xl"
  />
</div>
`;
